src/nvim/ex_cmds.c

```diff
@@ -3510,19 +3585,26 @@ void do_sub(exarg_T *eap)
         /*
          * 3. substitute the string.
          */
-        if (do_count) {
-          /* prevent accidentally changing the buffer by a function */
+        if (subflags.do_count) {
+          // prevent accidentally changing the buffer by a function
           save_ma = curbuf->b_p_ma;
-          curbuf->b_p_ma = FALSE;
+          curbuf->b_p_ma = false;
           sandbox++;
         }
-        /* get length of substitution part */
+        // Save flags for recursion.  They can change for e.g.
+        // :s/^/\=execute("s#^#gn")
+        subflags_T subflags_save = subflags;
+        // get length of substitution part
         sublen = vim_regsub_multi(&regmatch,
-            sub_firstlnum - regmatch.startpos[0].lnum,
-            sub, sub_firstline, FALSE, p_magic, TRUE);
-        if (do_count) {
+                                  sub_firstlnum - regmatch.startpos[0].lnum,
+                                  sub, sub_firstline, false, p_magic, true);
+        // Don't keep flags set by a recursive call
+        subflags = subflags_save;
```

Comment:
The same happens with Vim, and that's actually expected (now that I think it through).  The subflags changes are only partially related to the sandbox problem.  They're just ensuring that flags in recursive calls to `:substitute` don't affect other calls.

In the case of the test, `:s/^/\=execute("s#^##gn")/` causes `do_count` to still be in effect after the recursive `do_sub` call, which is what leads to the `sandbox--;` statement, so to truly test the fix you also need to remove the `if (sandbox > 0)` check.